### PR TITLE
oper: add support for system config load at app startup

### PR DESCRIPTION
**Pull request by adaraiseh**
Fixes part 1 #51

### CODE REVIEW

File: src/iproute2_sysrepo.c [hunk 1]
```diff
@@ -489,9 +500,49 @@ int ipr2_oper_get_items_cb(sr_session_ctx_t *session, uint32_t sub_id, const cha
...
+    char *output;
+    struct lyd_node *root_node = NULL;
+    fprintf(stdout, "%s: Started loading iproute2 running configuration.\n", __func__);
+    for (size_t i = 0; i < sizeof(ipr2_ip_modules) / sizeof(ipr2_ip_modules[0]); i++) {
+        fprintf(stdout, "%s: Loading module: %s data.\n", __func__, ipr2_ip_modules[i].module);
+        load_module_data(sr_session, ipr2_ip_modules[i].module, LYS_CONFIG_W, &root_node);
+    }
+
+    fprintf(stdout, "%s: Storing loaded data to sysrepo running datastore.\n", __func__);
+    ret = sr_edit_batch(sr_session, root_node, "merge");
```
Comment (Ali-aqrabawi) [inline src/iproute2_sysrepo.c:530]: any specific reason for having the edit as `merge` ? as this should be `replace` on startup/config_load, on startup we want to have the running DS identical to the linux config, and if we want, we can copy(merge) startup to running to push extra info

File: src/lib/oper_data.c [hunk 2]
```diff
@@ -182,37 +179,279 @@ int get_list_keys(const struct lysc_node_list *list, json_object *json_array_obj
...
-                    return EXIT_FAILURE;
+                    ret = EXIT_FAILURE;
+                    goto cleanup;
                 }
                 (*values_lengths)[key_count] = strlen(json_object_get_string(temp_value));
                 key_count++;
             } else {
                 // key value not found in json data.
-                return EXIT_FAILURE;
+                ret = EXIT_FAILURE;
```
Comment (Ali-aqrabawi) [inline src/lib/oper_data.c:197]: we are not freeing `key_values` and `values_lengths` here.

File: src/lib/oper_data.c [hunk 3]
```diff
@@ -182,37 +179,279 @@ int get_list_keys(const struct lysc_node_list *list, json_object *json_array_obj
...
+ * This function compairs schema node (s_node) parents names to lyd_node data tree node name, if the 
+ * schema node parent containers are not added to the lyd_node, this function adds them.
+ * 
+ * This function is to be used before processing leafs and leaf_lists to ensure their parent continers
+ * are added to the lyd data tree.
+ * 
+ * @param [in] s_node: schema node being processed (typically the s_node of a leaf or leaf_list).
+ * @param [in] parent_data_node: lyd data tree to which nodes under process are being attached to.
+ */
+void add_parent_node_if_needed(const struct lysc_node *s_node, struct lyd_node **parent_data_node)
```
Comment (Ali-aqrabawi) [inline src/lib/oper_data.c:265]: `add_missing_parents` might be a better name,

File: src/lib/oper_data.c [hunk 4]
```diff
@@ -182,37 +179,279 @@ int get_list_keys(const struct lysc_node_list *list, json_object *json_array_obj
...
+        add_parent_node_if_needed(s_node, parent_data_node);
+        if (json_object_is_type(temp_obj, json_type_array)) {
+            size_t n_json_flags = json_object_array_length(temp_obj);
+            for (size_t i = 0; i < n_json_flags; i++) {
+                struct json_object *inner_value = json_object_array_get_idx(temp_obj, i);
+                const char *value =
+                    map_value_if_needed(vmap_jobj, json_object_get_string(inner_value));
+                if (LY_SUCCESS !=
+                    lyd_new_term(*parent_data_node, NULL, s_node->name, value, 0, NULL)) {
+                    fprintf(stderr, "node %s creation failed.\n", s_node->name);
```
Comment (Ali-aqrabawi) [inline src/lib/oper_data.c:432]: can we add `__fun__ ` to the print

File: tests/cases/test_ip_netns.sh [hunk 5]
```diff
@@ -70,83 +70,5 @@ else
     exit 1
 fi
 
-
-
-# Exit with return value
```
Comment (Ali-aqrabawi) [inline tests/cases/test_ip_netns.sh:75]: can we undo this big change here, i know it might be line numbering, but please change only the line that you need to change, and we can fix spacing and other stuff in another commit.

**Review by Ali-aqrabawi — CHANGES_REQUESTED** [review]

File: src/lib/oper_data.c (hunk 2 above)
Comment (adaraiseh) [reply src/lib/oper_data.c:197]: Its the caller responsibility to free them

File: src/lib/oper_data.c (hunk 3 above)
Comment (adaraiseh) [reply src/lib/oper_data.c:265]: Ok

File: src/iproute2_sysrepo.c (hunk 1 above)
Comment (adaraiseh) [reply src/iproute2_sysrepo.c:530]: Good question, sysrepo normally copies startup_ds to running_ds meaning when we load iproute2-sysrepo it is expected to have some data there.
I didn't want to erase that config, until we decide how would final sequence of loading data would be, i.e.:
1-Load system config.
2-Apply sysrepo startup config.
3-subscribe to config changes.

Suppose we agree on this sequence, then:
If we do replace at 1, we will need to merge sysrepo startup config.
If we do merge at 1, we just apply startup config without any further operation.

Bottom line, this is up to change as we add support for sysrepo startup config load.

**Comment by adaraiseh** [comment]
code structure was built to support operational data load only, its getting really complicated trying to adapt it to loading config data.
working on complete code refactor.

File: src/iproute2_sysrepo.c (hunk 1 above)
Comment (adaraiseh) [reply src/iproute2_sysrepo.c:529]: changed to replace, we will open up this subject when adding sysrepo startup config load.

File: tests/cases/test_ip_netns.sh (hunk 5 above)
Comment (adaraiseh) [reply tests/cases/test_ip_netns.sh:75]: its not just a few lines, the whole scrip is duplicated ;)
I reverted the change, we do cleanup separately as suggested.
